keps/sig-cluster-lifecycle/20190122-Certificates-copy-for-kubeadm-join--control-plane.md

```diff
@@ -0,0 +1,330 @@
+   deleted by the `TokenCleaner` controller
+
+6. RBAC rules will be created to ensure access to the above config map to users
+   in the `system:node-bootstrapper` group (the bootstrap tokens).
+
+**On control plane node 2:**
+
+```bash
+kubeadm kubeadm join --control-plane --certificate-key={key from step above}
+```
+
+The new `--certificate-key` will trigger following actions:
+
+1. The `kubeadm-certs` config will be retrived from the cluster
+2. Cluster certificates will be decrypted using the provided key, and
+   hen stored on the disk
+
+### Implementation Details
+
+#### The encryption key
+
+The encryption key will use the same format as Bootstrap Tokens.
+See [Authenticating with Bootstrap Tokens](https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/#token-format)
+for more details.
+
+The default lifecycle of the encryption key will be the following:
+
+- a random encryption-key will be created by `kubeadm init --upload-certs`
+- the encryption-key will be printed in the output of the `kubeadm init --upload-certs`
+  command, but never stored in the cluster
+- the encryption-key will expire/become useless as soon as the `kubeadm-certs` will
+  be automatically deleted
+
+Following variant of the default lifecycle will be supported:
+
+1. The user will be allowed to pre-generate an encryption-key and pass it to
+   `kubeadm init --upload-certs` command using the kubeadm config file
+2. The user will be allowed to hide the encryption-key from the kubeadm output
+   using the `skip-token-print` or (a similar flag)
+3. The user will be allowed to pass the encryption-key to `kubeadm join --control-plane`
+   command using the kubeadm config file instead of the `--certificate-key` flag
+4. The user will be allowed to generate encryption-key after the first one expires by
+   invoking the `kubeadm init phases upload-certs` command (the command will
+  re-create/override the existing `kubeadm-certs` secret as well)
+
+#### The kubeadm-certs secret
+
+The `kubeadm-certs` secret will be stored in the `kube-system` namespace; RBAC rules
+for ensuring access to the users in the `system:node-bootstrapper` group
+(the bootstrap tokens) will be created.
+
+The `kubeadm-certs` secret `ownerReference` attribute will be set to the TTL-token
+UID as described in the example below; See [Garbage Collection](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/)
+for more details.
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: kubeadm-certs
+  namespace: kube-system
+  ownerReferences:
+  - apiVersion: v1
+    controller: true
+    blockOwnerDeletion: true
+    kind: Secret
+    name: bootstrap-token-abcdef
+    uid: 84e913dc-191a-11e9-96ae-0242f550e41f
+type: Opaque
+data:
+  ...
+```
+
+The `kubeadm-certs` secrets will contain the cluster certificates encrypted
+using TBD algorithm / the encryption key described above, and then base64
+encoded as usual.
```

Comment:
we never spoke about the algorithm.
should this be a technical detail outside of the scope of the KEP?
in my tests i did AES-256 in Galois/Counter Mode (GCM).

Reply:
If we really wanted to we could wrap the algorithm in an interface so we can swap it out if it ends up being insecure (would that even be mathematically possible?). I think at this point it's an implementation detail outside the scope of this KEP.

Reply:
agreed